src/components/survey/other-structure.tsx

```diff
@@ -0,0 +1,308 @@
+import { Stack, Button, Typography, TextField, Accordion, AccordionSummary, AccordionDetails, Box, List } from "@mui/material";
+import Api from "api";
+import { useAppSelector } from "app/hooks";
+import { ErrorContext } from "components/error-handler/error-handler";
+import { selectKeycloak } from "features/auth-slice";
+import { Building, OtherStructure } from "generated/client";
+import strings from "localization/strings";
+import * as React from "react";
+import GenericDialog from "components/generic/generic-dialog";
+import WithDebounce from "components/generic/with-debounce";
+import { Add, ExpandMore } from "@mui/icons-material";
+
+/**
+ * Component properties
+ */
+interface Props {
+  surveyId: string;
+}
+
+/**
+ * Component for other structures
+ */
+const OtherStructures: React.FC<Props> = ({ surveyId }) => {
+  const keycloak = useAppSelector(selectKeycloak);
+  const errorContext = React.useContext(ErrorContext);
+  const [ building, setBuilding ] = React.useState<Building>();
+  const [ creatingOtherStructure, setCreatingOtherStructure ] = React.useState(false);
+  const [ deletingOtherStructureId, setDeletingOtherStructureId ] = React.useState<number>();
+  const [ newOtherStructureName, setNewOtherStructureName ] = React.useState("");
+  const [ newOtherStructureDesc, setNewOtherStructureDesc ] = React.useState("");
+
+  /**
+   * Fetch building
+   */
+  const fetchBuilding = async () => {
+    if (!keycloak?.token || !surveyId) {
+      return;
+    }
+
+    try {
+      const fetchedBuildings = await Api.getBuildingsApi(keycloak.token).listBuildings({
+        surveyId: surveyId
+      });
+
+      if (fetchedBuildings.length !== 1) {
+        return;
+      }
+
+      setBuilding(fetchedBuildings[0]);
+    } catch (error) {
+      errorContext.setError(strings.errorHandling.buildings.list, error);
+    }
+  };
+
+  React.useEffect(() => {
+    fetchBuilding();
+  }, []);
+
+  /**
+   * Updates building
+   *
+   * @param updatedOwnerInformation updated owner information
+   */
+  const updateBuilding = async (updatedOwnerInformation: Building) => {
+    if (!keycloak?.token || !building?.id) {
+      return;
+    }
+
+    try {
+      Api.getBuildingsApi(keycloak.token).updateBuilding({
+        surveyId: surveyId,
+        buildingId: building.id,
+        building: updatedOwnerInformation
+      });
+    } catch (error) {
+      errorContext.setError(strings.errorHandling.buildings.update, error);
+    }
+  };
+
+  /**
+   * Delete other structure confirm handler
+   */
+  const onDeleteOtherStructureConfirm = async () => {
+    if (!building?.otherStructures || deletingOtherStructureId === undefined) {
+      return;
+    }
+
+    const updatedOtherStructures = building.otherStructures;
+
+    updatedOtherStructures.splice(deletingOtherStructureId, 1);
```

Comment:
This will probably create some unwanted state update without calling the set state function. I'd suggest that you use immer library if you need to update deep state structures